### PR TITLE
Rss ecomm 02 09

### PR DESCRIPTION
**Pull request by JoMarianek**
1. Task: [link](https://github.com/rolling-scopes-school/tasks/blob/master/tasks/eCommerce-Application/Sprints/Sprint2/RSS-ECOMM-2_09.md)
2. Screenshot: 
image

3. Change description: implemented registration page with required validation and error messages. I haven't implemented responsive design for small screens.
4. Change rationale

### CODE REVIEW

File: package.json
Comment (vetalapo) [inline package.json:1]: Please revert changes to this file.

File: src/data/countries.ts
Comment (vetalapo) [inline src/data/countries.ts:1]: We need our own source of countries. Please remove this file.

File: src/validation/registValidationSchema.ts
Comment (vetalapo) [inline src/validation/registValidationSchema.ts:1]: Please remove this file. We need only what vuetify provides. No need to complicates things.

File: src/views/RegistrationView.vue [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+            country: '',
+            password: '',
+            loading: false,
+            
+            rules: {
+                required: (value: string) => !!value || "Required",
+                noSpeacialChar: (value: string) => /^[a-zA-Z\s]*$/.test(value) || "No special characters allowed",
+                minLength: (minLength: number, message: string) => (value: string) => value.length >= minLength || message,
+                email: (value: string) => /^[--9^-~A-Z!#-'*+=?]+@[a-z0-9A-Z](?:[a-z0-9A-Z-]{0,61}[a-z0-9A-Z]|)(?:\.[a-z0-9A-Z](?:[a-z0-9A-Z-]{0,61}[a-z0-9A-Z]|))*$/.test(value) || "Invalid Email address",
+                minEight: (value: string) => value.length >= 8 || "Min 8 characters",
```
Comment (vetalapo) [inline src/views/RegistrationView.vue:23]: Why there is two "minEight" and "minLength"? Please leave just one.

File: src/views/RegistrationView.vue [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+            lastName: '',
+            streetName: '',
+            city: '',
+            email: '',
+            zip: '',
+            country: '',
+            password: '',
+            loading: false,
+            
+            rules: {
```
Comment (vetalapo) [inline src/views/RegistrationView.vue:18]: We have a lot of duplicating rules for registration and login pages. Maybe we could extract these rules in a separate class and import them?

File: src/views/RegistrationView.vue [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+                if (!value) return "Required";
+                const today = new Date();
+                const birthDate = new Date(value);
+                let age = today.getFullYear() - birthDate.getFullYear();
+                const monthDifference = today.getMonth() - birthDate.getMonth();
+                if (monthDifference < 0 || (monthDifference === 0 && today.getDate() < birthDate.getDate())) {
+                    age--;
+                }
+                return age >= 13 || "You must be at least 13 years old.";
+                }
```
Comment (vetalapo) [inline src/views/RegistrationView.vue:40]: Starting line 30 please make indents and spaces between blocks so it is easier to read. Maybe extracting it to a function would be better?

File: src/views/RegistrationView.vue [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    }
+</script>
+
+
+<template>
+    <v-layout>
+        <v-main id="registration-container" class="d-sm-flex align-center justify-center ga-16">
+            <div class="text-container d-flex flex-column align-center text-center mb-8">
+                <h1 class="text-center">Join us<br>today!</h1>
+                <h2 class="text-center pa-4" :style="{ color: '#099A9A' }">Sign up to get access to exclusive features and benefits.</h2>
```
Comment (vetalapo) [inline src/views/RegistrationView.vue:63]: Please don't assign styles like this. :style in Vue is for reactive changes. What you did could be put just in styles section.

File: src/views/RegistrationView.vue [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+                rounded="lg"
+            >
+
+                <v-form v-model="form" @submit.prevent="onSubmit">
+                    <div class="app-logo text-center d-sm-flex justify-center ga-5">
+                        <h2 class="brand">eCommerce</h2><span class="app">app</span>
+                    </div>
+
+                    <div >
+                        <h3 class="text-center pa-6" :style="{ color: '#099A9A' }">Sign up to get goods <br><span :style="{ fontWeight: 300}">you love</span></h3>
```
Comment (vetalapo) [inline src/views/RegistrationView.vue:78]: Please don't assign styles like this. :style in Vue is for reactive changes. What you did could be put just in styles section.

File: src/views/RegistrationView.vue [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        </v-main>
+    </v-layout>
+</template>
+
+<style scoped>
+    #registration-container {
+        background: 
+        linear-gradient(rgba(0, 0, 0, 0.5), rgba(0, 0, 0, 0.5)),
+         url("../assets/img/eCommerce-registration.svg") left no-repeat,
+         url("../assets/img/background.png") center/cover no-repeat;
```
Comment (vetalapo) [inline src/views/RegistrationView.vue:247]: Please indent with 4 white spaces.

File: src/views/RegistrationView.vue [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+                        block
+                    >
+                        Create an account
+                    </v-btn>
+
+                    <v-card-text class="text-center">
+                        <p>Already have an account?</p>
+                        <a
+                            class="text-blue text-decoration-none"
+                            href="#"
```
Comment (vetalapo) [inline src/views/RegistrationView.vue:229]: Please put "/login" in href. We already have the login page.

File: package.json [hunk 8]
```diff
@@ -18,14 +18,17 @@
...
+        "yup": "^1.4.0"
     },
     "devDependencies": {
         "@rushstack/eslint-patch": "^1.8.0",
         "@tsconfig/node20": "^20.1.4",
+        "@types/country-list": "^2.1.4",
         "@types/jsdom": "^21.1.6",
         "@types/node": "^20.12.5",
         "@vitejs/plugin-vue": "^5.0.4",
+        "@vue/cli-plugin-typescript": "^5.0.8",
```
Comment (vetalapo) [inline package.json:31]: If you including the plugin, I think you'll also have write rules for it.

File: src/assets/img/background.png
Comment (vetalapo) [inline src/assets/img/background.png:1]: It looks like left and bottom have empty areas? Is it per design?

File: src/views/RegistrationView.vue [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+            menu: false,
+            birthDate: null,
+            firstName: '',
+            lastName: '',
+            streetName: '',
+            city: '',
+            email: '',
+            zip: '',
+            country: '',
+            password: '',
```
Comment (vetalapo) [inline src/views/RegistrationView.vue:15]: Please use double quotes "" for values

**Review by vetalapo — CHANGES_REQUESTED** [review]

**Review by vetalapo — CHANGES_REQUESTED** [review]
Please add screenshot of the page in the PR

File: package.json (hunk 8 above)
Comment (JoMarianek) [reply package.json]: Ok, I removed it

**Comment by JoMarianek** [comment]
@vetalapo Implemented all requested changes, could you please take a look and approve?

**Comment by JoMarianek** [comment]
@vetalapo implemented changes

File: src/views/LoginView.vue [hunk 10]
```diff
@@ -1,37 +1,13 @@
...
-                minOneLowerCase: (value: string) =>
-                    /(?=.*[a-z])/.test(value) || "Must contain at least one lowercase letter",
-                minOneUpperCase: (value: string) =>
-                    /(?=.*[A-Z])/.test(value) || "Must contain at least one UPPERCASE letter",
-                minOneSpecialChar: (value: string) =>
-                    /(?=.*\W)/.test(value) || "Must contain at least one special character",
-                noLeadingTrailingWhitespace: (value: string) =>
-                    /^(?! ).*[^ ]$/.test(value) || "Must not contain leading or trailing whitespace"
-            }
+            commonRules: ValidationRules,
```
Comment (vetalapo) [inline src/views/LoginView.vue:10]: Missed this, please don't leave trailing commas.

File: src/views/RegistrationView.vue [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+                    const today = new Date();
+                    const birthDate = new Date(value);
+                    let age = today.getFullYear() - birthDate.getFullYear();
+                    const monthDifference = today.getMonth() - birthDate.getMonth();
+
+                    if (monthDifference < 0 || (monthDifference === 0 && today.getDate() < birthDate.getDate())) {
+                        age--;
+                    }
+
+                return age >= 13 || "You must be at least 13 years old.";
```
Comment (vetalapo) [inline src/views/RegistrationView.vue:38]: Please indent with 4 spaces this too.

**Review by vetalapo — CHANGES_REQUESTED** [review]
Please take a look at the last couple of things. It is almost ready to merg.

**Comment by JoMarianek** [comment]
@vetalapo implemented changes

**Review by vetalapo — APPROVED** [review]

**Pull request merged**